tests/optim/test_optim.py

```diff
@@ -231,3 +234,56 @@ def model():
         optimizer({param})
 
     assert_equal(param, param.new_zeros(param.shape), prec=1e-2)
+
+
+def test_name_preserved_by_to_pyro_module():
+    features = torch.randn(4)
+    data = torch.randn(3)
+
+    class Model(PyroModule):
+        def __init__(self):
+            super().__init__()
+            self.scale = PyroParam(torch.ones(3), constraints.positive)
+            self.loc = torch.nn.Linear(4, 3)
+
+        def forward(self, features, data):
+            loc = self.loc(features)
+            scale = self.scale
+            with pyro.plate("data", len(data)):
+                pyro.sample("obs", dist.Normal(loc, scale),
+                            obs=data)
+
+    model = Model()
+    params = list(model.parameters())
+    param_names = set()
+
+    def optim_config(param_name):
+        param_names.add(param_name)
+        return {"lr": 0.0}
+
+    # Record while model.loc is an nn.Module.
+    loss = poutine.trace(model).get_trace(features, data).log_prob_sum()
+    loss.backward()
+    adam = optim.Adam(optim_config)
+    adam(params)
+    assert param_names
+    expected_param_names = param_names.copy()
+    del adam, loss
+    param_names.clear()
+    pyro.clear_param_store()
+
+    # Record while model.loc is a PyroModule.
+    to_pyro_module_(model.loc)
+    loss = poutine.trace(model).get_trace(features, data).log_prob_sum()
+    loss.backward()
+    adam = optim.Adam(optim_config)
+    adam(params)
+    assert param_names
+    actual_param_names = param_names.copy()
+    del adam, loss
+    param_names.clear()
+    pyro.clear_param_store()
+
+    print(expected_param_names)
+    print(actual_param_names)
+    assert actual_param_names == expected_param_names
```

Comment:
I saw some simplifications but still couldn't figure out what are the expected names here. Could you make this assertation more verbose like
```
assert acutal_param_names == ...
assert actual_param_names == expected_param_names
```

Reply:
Done.